File: 300.500/380.InsertDeleteGetRandomO1.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <stdlib.h>
#include <time.h>
using namespace std;

class RandomizedSet {
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        srand(time(nullptr));
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if(data.find(val) != data.end())
            return false;
        data.insert(val);
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        auto it = data.find(val);
        if(it == data.end())
            return false;
        data.erase(it);
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        int pos = rand()%data.size();
        auto it = data.begin();
        advance(it, pos);
        return *it;
    }
private:
    unordered_set<int> data;
};

int main(int argc, char const *argv[]){
    RandomizedSet cls;
    cout << boolalpha << cls.insert(1) << " -> true" << endl;
    cout << boolalpha << cls.remove(2) << " -> false" << endl;
    cout << boolalpha << cls.insert(2) << " -> true" << endl;
    cout << cls.getRandom() << endl;
    cout << boolalpha << cls.remove(1) << " -> true" << endl;
    cout << boolalpha << cls.insert(2) << " -> false" << endl;
    cout << cls.getRandom() << endl;
    return 0;
}

File: go/300.500/380.InsertDeleteGetRandomO1.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type RandomizedSet struct {
	kv   map[int]int
	data []int
}

/** Initialize your data structure here. */
func Constructor() RandomizedSet {
	rand.Seed(time.Now().UnixNano())
	return RandomizedSet{
		kv:   map[int]int{},
		data: []int{},
	}
}

/** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
func (this *RandomizedSet) Insert(val int) bool {
	if _, ok := this.kv[val]; ok {
		return false
	}
	this.kv[val] = len(this.data)
	this.data = append(this.data, val)
	return true
}

/** Removes a value from the set. Returns true if the set contained the specified element. */
func (this *RandomizedSet) Remove(val int) bool {
	if pos, ok := this.kv[val]; ok {
		//将要删除的元素移动到末尾
		end := len(this.data) - 1
		if end != pos {
			this.data[pos] = this.data[end]
			this.kv[this.data[end]] = pos
		}
		this.data = this.data[:end]
		delete(this.kv, val)
		return true
	}
	return false
}

/** Get a random element from the set. */
func (this *RandomizedSet) GetRandom() int {
	pos := rand.Intn(len(this.data))
	return this.data[pos]
}

func main() {
	cls := Constructor()
	fmt.Println(cls.Insert(1), true)
	fmt.Println(cls.Remove(2), false)
	fmt.Println(cls.Insert(2), true)
	fmt.Println(cls.GetRandom())
	fmt.Println(cls.Remove(1), true)
	fmt.Println(cls.Insert(2), false)
	fmt.Println(cls.GetRandom())
}
